packages/react-component-library/src/components/NumberInputE/NumberInputE.tsx

```diff
@@ -190,6 +191,60 @@ export const NumberInputE: React.FC<NumberInputEProps> = ({
     )
   }
 
+  /**
+   * Block invalid input before the DOM is updated where possible (to
+   * avoid the cursor jumping).
+   *
+   * Note: React simulates this event using other native events.
+   */
+  const handleBeforeInput = (
+    event:
+      | React.FormEvent<HTMLInputElement>
+      | React.CompositionEvent<HTMLInputElement>
+  ) => {
+    // Unfortunately, the event seems to be typed incorrectly at present
+    // (without the data property)
+    if (!('data' in event)) {
+      return
+    }
+
+    // data could contain a single character, or something longer
+    // in case of pasting. The text could be an insertion or a
+    // replacement (if existing text was selected).
+    const { data } = event
+
+    if (!areCharactersValid(data)) {
+      event.preventDefault()
+      event.stopPropagation()
+    }
+  }
+
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (!isValueValid(event.currentTarget.value)) {
+      return
+    }
+
+    const newValue = getNewValue(event)
+
+    if (!newValue || canCommit(newValue)) {
+      setCommittedValue(newValue)
```

Comment:
Maybe because this interacts with `setCommittedValue` it is an indication that it could belong in a hook.